### Q10
solution from Chaolun

#### 1
Define:
$$
x^T=[x_1, x_2,...,x_n]
$$
$$
e^T=[1, 1,...,1]\space\space\space\space\space\space(e\space has\space the\space length\space m1)
$$
$$
w^T=[w_1, w_2,...,w_{m1}]
$$
$$
T=\begin{vmatrix}
-t_1^T-\\
-t_2^T-\\
\vdots \\
-t_{m1}^T-\\
\end{vmatrix}
$$
Also define the following operator as elementary wise matrix operation of multiply, taking squareroot, square and function evaluation:
$$*$$
$$\sqrt()$$
$$()^2$$
$$\Phi()$$
And the l2 norm was used to calculate the distance.

Then we can have:
$$
B=T-ex^T
$$
So the kth column of B is the difference between x and $t_k$

Then we can get:
$$
A=\Phi(\sqrt{B*Be})
$$
So the kth element of vector A is the activation of the kth hidden unit.

Finally we can get:
$$
y=A^Tw
$$
So the matrix equation is:
$$
y=[\Phi(\sqrt{(T-ex^T)^2e})]^Tw
$$
To calculate B, 2nm1 operations are required(for multiplication and substraction).

To calculate A, nm1+(n-1)m1+2m1 operations are required(for square, multiplication, taking squareroot and function evaluation).

To calculate y, 2m1-1 operations are required(for multiplication and addition).

So in total, there are 4nm1+3m1-1 operations required(O(nm1) complexity).

### 2
Define:
$$
x^T=[x_1, x_2,...,x_n]\space\space\space(n=10)
$$
$$
e^T=[1, 1,...,1]
$$
$$
w==\begin{vmatrix}
w_{11} & w_{12}....w_{1p}\\
w_{21} & w_{22}....w_{2p}\\
\vdots \\
w_{m1} & w_{m2}....w_{mp}\\
\end{vmatrix}
$$
where $w_{ij}$ means the weight between the ith hidden unit and the jth output unit
$$
T=\begin{vmatrix}
-t_1^T-\\
-t_2^T-\\
\vdots \\
-t_{m1}^T-\\
\end{vmatrix}
$$
Also define the following operator as elementary wise matrix operation of multiply, taking squareroot, square and function evaluation:
$$*$$
$$\sqrt()$$
$$()^2$$
$$\Phi()$$
And the l2 norm was used to calculate the distance.

Then we can have:
$$
B=T-ex^T
$$
So the kth column of B is the difference between x and $t_k$

Then we can get:
$$
A=\Phi(\sqrt{B*Be})
$$
So the kth element of vector A is the activation of the kth hidden unit.

Finally we can get:
$$
y=A^Tw
$$
So the kth element of vector y is the activation of the kth output unit.

So the matrix equation is:
$$
y=[\Phi(\sqrt{(T-ex^T)^2e})]^Tw
$$
To calculate B, 2nm operations are required(for multiplication and substraction).

To calculate A, nm+(n-1)m+2m operations are required(for square, multiplication, taking squareroot and function evaluation).

To calculate y, 2mp-p operations are required(for multiplication and addition).

So in total, there are 4nm+m+2mp-p operations required, which is 41m+2mp-p operations when n=10(O(mp) complexity).


In [73]:
import numpy as np
from timeit import default_timer as timer

def fi1(r, c):
    return 1.0/np.sqrt(r**2+c**2)

def fi2(r, c):
    return np.exp(-1*r**2/(2*c**2))

#initialization
n=10
p=5
m=10
c=1

x=np.random.uniform(0,1,[n,1])
e=np.ones([m,1])
T=np.random.uniform(0,1,[m,n])

# p=5
p=5
w=np.random.uniform(-1,1,[m,p])

#timing
start=timer()

#calculation
for _ in range(10000):
    B=T-np.dot(e,x.T)
    A=fi1(np.sqrt(np.dot(B*B,e)), c)
    y=np.dot(A.T,w)

#end timing
end=timer()

print("when p=%d, time consumed: %f, the output value is as following: "%(p, (end-start)))
print(y)

# p=10
p=10
w=np.random.uniform(-1,1,[m,p])

#timing
start=timer()

#calculation
for _ in range(10000):
    B=T-np.dot(e,x.T)
    A=fi1(np.sqrt(np.dot(B*B,e)), c)
    y=np.dot(A.T,w)

#end timing
end=timer()

print("when p=%d, time consumed: %f, the output value is as following: "%(p, (end-start)))
print(y)

# p=15
p=15
w=np.random.uniform(-1,1,[m,p])

#timing
start=timer()

#calculation
for _ in range(10000):
    B=T-np.dot(e,x.T)
    A=fi1(np.sqrt(np.dot(B*B,e)), c)
    y=np.dot(A.T,w)

#end timing
end=timer()

print("when p=%d, time consumed: %f, the output value is as following: "%(p, (end-start)))
print(y)

when p=5, time consumed: 0.116584, the output value is as following: 
[[ 0.14185822  0.13718936  1.26122439  0.40524829 -1.12042033]]
when p=10, time consumed: 0.121830, the output value is as following: 
[[-0.99411773  0.22394171 -1.49035449  0.7964547  -0.00845997 -0.67441762
   1.44657179  0.44868343 -0.35263324  0.73105679]]
when p=15, time consumed: 0.123300, the output value is as following: 
[[ 1.16269484 -1.29295402 -0.42553624 -2.2194504   4.70815477 -0.36782397
   2.489402    0.73666263 -0.91213885 -1.37320316  0.85655771  0.45469535
   0.19193    -0.08265677  0.26828006]]


In order to make the timing reliable, each calculation has been iterated 10000 times, the time consumed for p=5, 10 and 15 is 0.116584, 0.121830 and 0.123300 ($*10^{-4}$ s) separately. The answer could be different everytime running the code. The scaling of time with respect to p is about $6*10^{-8} $

#### 3
Define:
$$
x^T=[x_1, x_2,...,x_n]\space\space\space(n=10)
$$
$$
e^T=[1, 1,...,1]
$$
$$
w==\begin{vmatrix}
w_{11} & w_{12}....w_{1p}\\
w_{21} & w_{22}....w_{2p}\\
\vdots \\
w_{m1} & w_{m2}....w_{mp}\\
\end{vmatrix}
$$
$$
T=\begin{vmatrix}
-t_1^T-\\
-t_2^T-\\
\vdots \\
-t_{m1}^T-\\
\end{vmatrix}
$$
It can also be written as matrix equation as:
$$
y=[\Phi(\sqrt{(T-ex^T)^2e})]^Tw
$$
where:
$$
\Phi(r)=\frac{1}{1+exp(-r)}
$$
Similar to subquestion 1 and 2, $\Phi()$ is the elementary wise operation over matrix

#### 4
Given that the insertion and deletion operator are not required for this data structure and the index of element can maintain unchanged during calculation, a simple 2D array will be good enough for this purpose. Since there is a large number of radial basis fields, it may have problem store all of the data in a continuous block(although it will be the most efficient implementation given enough memory). So the 2D array can be implemented as "array of pointers", with each pointer pointed to a 1D array(dynamically allocated in heap), which represents the position of the neuron. The indexing of array can be done in O(1), which makes the data accessing speed of 2D array much faster than other complicate datastructure(linked list, hash table for example). Here is an example of the implication (in c++):


//declearation in .h file

class TwoDimArray{

     private:

           int* pArr;
   
           int rows;
   
           int cols;

     public:

           TwoDimArray(int rows, int cols);
   
           int* operator[](int row);
   
           ~TwoDimArray();

};


//implementation .cpp file

TwoDimArray::TwoDimArray(int nrOfRows, int nrOfCols){

       rows = nrOfRows;
   
       cols = nrOfCols;

       //allocate memory
   
       pArr = new int[rows * cols];
   
}


int* TwoDimArray::operator [](int row){

       return &pArr[row * cols];
   
}


TwoDimArray::~TwoDimArray(){

       delete[] pArr;
   
}